Redesign

athanclark opened the pull request:
An absolute ton went into this revision. Here is a brief summary:

# Testing

There's a much more thorough test suite, first off - the `TypedArray` mechanism was most important - to prove that all generated values, for the different bit-caste sizes, operate exactly as expected for all operations, was a critical milestone.

# Foreign Interface

The previous version relied on `Uint8Arrays` as the fundamental representation in JavaScript for most types - that is no longer the case; the actual JavaScript `DataView`, `ArrayBuffer`, and `TypedArray` types are the underpinning representation for the PureScript types.

# Class-based API

Because of the commonality between APIs among the different `TypedArray` types, I took the liberty of designing it with classes and instances in-mind. The performance overhead should be negligible, especially when using [rollup-plugin-purs](https://github.com/Pauan/rollup-plugin-purs).

There's also a classy interface for `DataViews`, to choose the getter and setters for each binary representation.

# No more Data.ArrayBuffer.Show, No more String encodings

As @AlexaDeWit mentioned in #20, string encodings will be done with `purescript-arraybuffer-codecs`. The `Show` module used a foreign npm module, `inspect` I think, and it just felt gnarly, so I got rid of it.

# Generators

`MonadGen` values live in `Data.ArrayBuffer.*.Gen`, for unit test generation.

# Type-Level Kit

There's some extra mappings in `Data.ArrayBuffer.ValueMapping`, where byte length and application-level type casts are defined for each `ArrayViewType`. This was necessary to keep `DataView` and `TypedArray`s consistent at compile-time.

# Value-Level Kit

The TypedArray interface at `Data.ArrayBuffer.Typed` provides all the tools you'd expect from an iterable structure - folds, maps, traversals, searching, querying, filtering, slicing, etc. are all in the form of [Snoyman's mono-traversable](http://hackage.haskell.org/package/mono-traversable).


That's about it! Please let me know what you think, any feedback is helpful.

athanclark committed:
ripping out untrue functionality of arraybuffers

athanclark committed:
rename to stay true to implementation

athanclark committed:
converting to Effect.Uncurried, using exception catching, naming
anonymous foreign functions for stack traces

athanclark committed:
module documentation

athanclark committed:
preparing for major redesign of typed interface - supporting all
functions

athanclark committed:
polyfill

athanclark committed:
module doc

athanclark committed:
using `new` as binder - also, remainder function

athanclark committed:
copyWithin

athanclark committed:
all, any ~ every, some

athanclark committed:
fill semantics

athanclark committed:
toString

athanclark committed:
subarray

athanclark committed:
in-place sort

athanclark committed:
slice

athanclark committed:
moving `set` around a little

athanclark committed:
reverse in-place

athanclark committed:
typed map

athanclark committed:
joining an array

athanclark committed:
filter

athanclark committed:
toArray typed properly

athanclark committed:
folds

athanclark committed:
elem, need to do find and indexing

athanclark committed:
better arraybuffer slice, more stuff

athanclark committed:
the uint library stores values greater than 2^31 as a negative integer -
this is not a correct implementation currently in DataView

athanclark committed:
index of

athanclark committed:
more instances

athanclark committed:
honestly I forgot about floating points

athanclark committed:
the binary serialization of javascript Numbers for TypedArrays follows
the intended values that would be expected in javascript - negative
encodings for proper Uint32 is invalid in this case. Because of this,
Uint32s are represented in the positive domain of 64bit floats (js
Number)

AlexaDeWit committed:
Remove text encoding dependency

AlexaDeWit committed:
Remove show and update readme

athanclark committed:
generation functions - need to generate a bona-fide float32

athanclark committed:
arbitrary floating point values

athanclark committed:
more correct

athanclark committed:
more generators

athanclark committed:
simple byteLength / bytesPerValue === length test

athanclark committed:
fromArray <<< toArray iso

athanclark committed:
better docs

athanclark committed:
cosmetic

athanclark committed:
using uint

athanclark committed:
fill y xs => all (== y) xs

athanclark committed:
generic instance of `WithOffset` structure

athanclark committed:
cosmetic, with offset

athanclark committed:
removed funky show module

athanclark committed:
set singleton is eq

athanclark committed:
all implies any

athanclark committed:
filter, all / any

athanclark committed:
elem with unsafeAt

athanclark committed:
any implies find

athanclark committed:
findIndex implies at

athanclark committed:
indexOf implies at

athanclark committed:
lastIndexOf imples at

athanclark committed:
quickcheck combinators

athanclark committed:
cosmetic

Commit:
sort is idempotent, reverse tests

Commit:
sort is correct

Commit:
toString

Commit:
dual subArray and slice references

Commit:
really weird test results - subArray isn't mutable when supplied with a nonzero argument?

Commit:
peculiar behavior of subArray

Commit:
verifying the mutability of whole subArrays

Commit:
better docs

Commit:
misimplementation

Commit:
copy within self

Commit:
copyWithin verified

Commit:
more verification for copyWithin

Commit:
gen arraybuffer and dataview

Commit:
more qualified imports

Commit:
using inline ternary op

Commit:
different gen paradigm

Commit:
moving classes around - safe dataView modification

athanclark committed:
generating a data view with an offset, for a particular value

athanclark committed:
properly using uint

athanclark committed:
runners

athanclark committed:
using type-declared value sizes

athanclark committed:
verified data views

athanclark committed:
cosmetic

Commit:
better docs, noting how subArray / new use Data.ArrayBuffer.ArrayBuffer.slice

Commit:
reformatting arguments

Commit:
final touch-ups

Commit:
removing obsolete tests

athanclark commented:
Ping @jacereda - is there anything you'd like me to fix?

jacereda commented:
I’m in the process of upgrading my laptop and didn’t have time to check yet, but I guess the docs shouldn’t be there, since they are generated when uploading to Pursuit.

> El 18 dic 2018, a las 17:56, Athan Clark <notifications@github.com> escribió:
> 
> Ping @jacereda - is there anything you'd like me to fix?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

Commit:
removed generated docs

athanclark commented:
You got it!

jacereda commented on src/Data/ArrayBuffer/Typed.js:
```diff
@@ -1,75 +1,219 @@
+
+// Typed Arrays
+
+exports.newUint8ClampedArray = function newUint8ClampedArray (a,mb,mc) {
```
What about something like:

```
function newArray(f, a, mb, mc) {
    return mc === null ? ( mb === null ? new f(a)
                                       : new f(a,mb)
                         )
                       : new f(a,mb,mc);
}

exports.newUint8ClampedArray = function newUint8ClampedArray (a,mb,mc) {
  return newArray(Uint8ClampedArray, a, mb, mc);
};

```

athanclark replied:
Oh jeez, sorry. I thought JavaScript would be context-sensitive with `new`, and attempt to find a JS object constructor `f` or something... Shoot. Here let me fix this, one sec.

jacereda commented on src/Data/ArrayBuffer/DataView.js:
```diff
@@ -3,43 +3,35 @@
+exports.getterImpl = function getterImpl (s, l, e, v, o) {
+    return v[s].call(v,o,e);
+};
+
```
The old approach (returning Maybe) is 3 times faster.

athanclark replied:
You're sure? Is it because of the imports, or something more fundamental? I'd be happy to make a bench suite, there's some nice libs out there.

jacereda replied:
I guess it's just that `try` is expensive.

This is what I used:

```

module Test.Benchmark where

import Prelude

import Benchmark (fn, runBench)
import Data.ArrayBuffer.DataView (getter, getter', whole)
import Data.ArrayBuffer.Typed (buffer, fromArray)
import Data.ArrayBuffer.Types (ArrayView, Float32)
import Effect (Effect)

benchmark :: Effect Unit
benchmark = runBench do
  let av :: ArrayView Float32
      av = fromArray [1.0,2.0,4.0]
      dv = whole (buffer av)
  fn "goodgetter" (\x -> getter "Float32" 1 true dv x) 0
  fn "goodgetter'" (\x -> getter' "Float32" 1 true dv x) 0
  fn "badgetter" (\x -> getter "Float32" 1 true dv x) 100
  fn "badgetter'" (\x -> getter' "Float32" 1 true dv x) 100

```


I used my fork of `purescript-benchmark` for benchmarking, there's a pending PR on that:

    "purescript-benchmark": "jacereda/purescript-benchmark-1#master"

athanclark replied:
Awesome. Sorry I've taken so long - I got sidetracked with another project. I'll try and have my edition fleshed out with a total benchmark suite, with competing metrics for these implementations you've alluded to here, hopefully out before Jan. 7th. Thanks for being so patient!

athanclark replied:
I'm just going to take your word for this one :) I've gotten way too sidetracked to invest in a proper benchmark suite, but I'll eventually get one fleshed out. I'd just rather have something for everyone to use for now.

athanclark committed:
omitting `try` in DataView.js

athanclark committed:
Merge branch 'redesign' of github.com:athanclark/purescript-arraybuffer into redesign

athanclark committed:
better code re-use

athanclark commented:
@jacereda Alright, sorry for the wait. I wasn't able to put together a benchmarking suite yet, but I've submitted changes that reflected your comments above. Sorry again this has taken so long :x

jacereda commented on src/Data/ArrayBuffer/DataView.js:
```diff
@@ -3,43 +3,35 @@
+};
+
+exports.getterImpl = function getterImpl (data, s, l, e, v, o) {
+    return ((o + 1) >>> 0) <= v.byteLength ? data.just (v[s].call(v,o,e)) : data.nothing;
```
I think that should read `o + l`, not `o + 1`.

athanclark replied:
Holy Toledo... Sorry. I'll fix that right away.

jacereda commented on src/Data/ArrayBuffer/DataView.js:
```diff
@@ -3,43 +3,35 @@
+    return ((o + 1) >>> 0) <= v.byteLength ? data.just (v[s].call(v,o,e)) : data.nothing;
+};
+
+exports.setterImpl = function setterImpl (s,e,v,n,o) {
```
I don't know why this didn't perform range-checking in the same way `getterImpl` does... the previous version was probably broken, we could fix it here now...

athanclark replied:
I also think I got a little carried away with the purescript class - I think each instance should be just for an `ArrayBufferType`, and have `getBE/LE`, `setBE/LE` for each type.

Should this mean that `set*` should return `Effect Boolean` for success / failure?

jacereda replied:
I think so, that's what -arrays does:

https://pursuit.purescript.org/packages/purescript-arrays/5.2.0/docs/Data.Array.ST#v:poke

athanclark replied:
Alright, I've got this implemented, at least I think how it should be.

Commit:
fixing implementation

Commit:
reorganized setter

Commit:
better design

jacereda commented:
I'll try to fix https://github.com/NixOS/nixpkgs/issues/53597 and after that I'll hopefully devote some time to merge this in the next couple of days.

athanclark commented:
No worries. I'm making a few more edits to the PR also - I needed to move stuff around in purescript-uint, so things might still be a while off.
Thanks!

Commit:
omitting Arbitrary constraint

Commit:
cosmetic

athanclark commented:
Ping :) sorry to pester

jacereda committed:
Merge v9-redesign2

jacereda committed:
Some cosmetic changes and fixes.

Fix limits in generators.
Make limits more explicit.
Simplify Float32 generator.

jacereda committed:
Sort includes

jacereda committed:
Sort includes

jacereda committed:
Still not happy with the names, renamed generators

jacereda commented:
The changes have been pulled to the v9 branch.

I've sorted imports, renamed generators and fixed a couple of errors in the limits. Still reviewing the API...

jacereda committed:
toNullable Nothing -> null

jacereda committed:
toNullable . Just - > notNull

jacereda committed:
Typo

jacereda commented on src/Data/ArrayBuffer/Typed.purs:
```diff
@@ -1,87 +1,544 @@
+-- | - `toArray` returns an array of numeric values
+class BinaryValue a t <= TypedArray (a :: ArrayViewType) (t :: Type) | a -> t where
+  -- | View mapping the whole `ArrayBuffer`.
+  whole :: ArrayBuffer -> ArrayView a
```
I guess we should have `whole`/`remainder`/`part` share the same return type. Did you keep it this way to be backwards-compatible?

jacereda replied:
I don't have formed an opinion yet, why should these be an `Effect`? Isn't this just like `slice` in `Array`? The signature for `slice` is `Int -> Int -> Array a -> Array a`...

athanclark replied:
I left `whole` to be pure, because it can't fail, but `remainder` and `part` can raise exceptions because of range errors. It's symmetric across the different modules, though.

jacereda commented on src/Data/ArrayBuffer/Typed.purs:
```diff
@@ -1,87 +1,544 @@
+  -- | Test a predicate to pass on all values
+  all :: (t -> Offset -> Effect Boolean) -> ArrayView a -> Effect Boolean
+  -- | Test a predicate to pass on any value
+  any :: (t -> Offset -> Effect Boolean) -> ArrayView a -> Effect Boolean
```
Why are the callbacks effectful here?

athanclark replied:
I was just trying to stay true to how the foreign js would operate; but you're totally right - effectful behaviour like this should be left to a traverseWithIndex.

Commit:
Cherry-pick 83370eb9f96aa1638f1c532d0d4c10bdad295d1c, Uint32 instances no longer use toNumber/fromNumber

Commit:
pure functions for filtering & querying

jacereda commented:
Please, sync with https://github.com/jacereda/purescript-arraybuffer/tree/v9 before additional commits or PR over that one.

jacereda committed:
Fix the tests

jacereda committed:
These share the implementation...

jacereda committed:
... and don't need to be part of the type class

jacereda committed:
These can also be removed from the type class

jacereda committed:
Simplify slice

jacereda committed:
Simplify subArray

athanclark commented:
Oh wow okay, let me get that squared up, sorry one second.

Commit:
Merge branch 'v9' of git://github.com/jacereda/purescript-arraybuffer into redesign

jacereda committed:
unsafeAt should be like unsafeIndex in arrays

jacereda committed:
Simplify fill

jacereda committed:
Better name

jacereda committed:
Use notNull

jacereda committed:
WIP, generate arrays with the size provided by QC, split functions that take ByteOffset

jacereda committed:
Cleanup

jacereda committed:
Fix set/setTyped

jacereda committed:
Cleanup

jacereda committed:
Fix allImpliesAnyTests

jacereda commented on src/Data/ArrayBuffer/DataView.purs:
```diff
@@ -67,113 +50,97 @@ foreign import byteOffset :: DataView -> ByteOffset
+  getBE :: AProxy a -> DataView -> ByteOffset -> Effect (Maybe t)
+  setBE :: AProxy a -> DataView -> t -> ByteOffset -> Effect Boolean
+  setLE :: AProxy a -> DataView -> t -> ByteOffset -> Effect Boolean
+
```
What's this? AFAIU a `DataView` is just a raw chunk of memory that you can access any way you want (peek/poke). It's what you would use to read from a raw buffer (think in binary serialisation protocols).

athanclark replied:
So technically, and this is as far as I understand too, ArrayBuffers are the lowest level interface. DataViews are just as you are saying - byte-indexable array with castable setting and getting - you can caste _how_ a value is set and received (via the `AProxy`), like a 32bit little-endian int or something. TypedArrays are just uniform DataViews for every observed value, and only in big-endian.

Is there something wrong with this binding you're seeing? I'm using the AProxy to help decide how much data to set / get.

jacereda replied:
Looks like I got confused reading the signatures, it's clear now.

I'll work on this again in a couple of days.

jacereda committed:
Use unsigned

jacereda committed:
Disable DataView tests

jacereda committed:
Simplifying

jacereda committed:
Introduce overAll1 to test over non-empty arrays

Commit:
using external float32 library

jacereda commented:
Oh, I didn't notice `fromNumber` was doing the conversion via `fround`, forget what I said.

athanclark commented:
Ahh shoot you're totally right, let me fix that.

athanclark commented:
Hmm, so here's a tricky case - the Bounded instance for Number just uses positive and negative infinity for it's bounds. Now, `Math.fround()` will return Infinity for `Math.fround(3.40283e38)`, but not for `3.40282e38`.

Should I just encode that behavior in the polyfill, then catch it into a `Maybe`? Either way, it shouldn't affect this PR.

athanclark commented:
There we go, float32 v0.1.0 should be satisfactory

jacereda commented:
> Hmm, so here's a tricky case - the Bounded instance for Number just uses positive and negative infinity for it's bounds. Now, `Math.fround()` will return Infinity for `Math.fround(3.40283e38)`, but not for `3.40282e38`.
> 

I wonder if `Bounded` should return the extreme non-infinite value, it looks more useful than `Infinite`s at least. Or maybe it should be decoupled in functions possibly returning infinite and functions returning extreme finite values.

jacereda committed:
WIP, making API more consistent with foldable-traversable methods

athanclark commented:
Alright, version v0.1.1 for float32 should meet this criteria.

jacereda commented:
> Alright, version v0.1.1 for float32 should meet this criteria.

I think the right solution would be another type class, I commented on that in this PR:

https://github.com/purescript/purescript-quickcheck/pull/97

But maybe it should be opened as a ticket in -prelude.

jacereda commented:
> I think the right solution would be another type class

And that's because the`Bounded` laws say that for any value, `bottom <= v <= top`, so even if it's more useful it's violating the law, because it fails for `Infinite`.

athanclark commented:
There are negative and positive infinites in Number, but either way I think this is derailing from arraybuffer and designating to float32. But I'll definitely integrate your code into the generic Data.Float32.Gen module, but I'm not sure about making a new Bounded class. Right now, the instance reflects the finite maximum and minimum values in the float32 range.

jacereda commented:
Did you notice this comment?

https://github.com/jacereda/purescript-arraybuffer/pull/23#discussion_r249243904

Are you working on an alternate API for that or should I just revert to the old one?

jacereda commented:
Looks like the old `DataView -> XXXArray` are no longer part of the API. How would you extract a `Int8Array` from an `ArrayBuffer` (a message received via web socket for instance) without performing copies?

jacereda committed:
Merge master

athanclark commented:
I was just going to leave that to the end-user - the fundamental data type for all of these is an ArrayBuffer, which can be obtained from any DataView or TypedArray via the `buffer` function - then you can make a new data view or typed array in the way you'd prefer - either `whole` for the entire buffer, or something like `remainder` or `part`, where you would only want access to a slice of the ArrayBuffer.

jacereda commented:
OK, makes sense.

jacereda committed:
Enabled DataView tests

jacereda committed:
Saturate arguments for runEffectFnX

jacereda committed:
Saturate arguments for run(Effect)FnX

jacereda commented on src/Data/ArrayBuffer/ArrayBuffer.purs:
```diff
@@ -1,52 +1,31 @@
 
 -- | Create an `ArrayBuffer` with the given capacity.
-foreign import create :: ByteLength -> Effect ArrayBuffer
+foreign import empty :: ByteLength -> ArrayBuffer
```
Do you think this should be an `Effect`?

jacereda replied:
I think it should, since:
```
let a = empty 16
foo a
bar a
```

Is quite different from

```
foo $ empty 16
bar $ empty 16
```

right?

athanclark replied:
Exactly :) they are shared references to the same mutable value. I feel like we should aim for a _pleasant_ median between a pure traversable structure, and an effectful mutable reference. But honestly I would be comfortable with either - we should just verbosely state in the documentation the semantic recognition of ArrayBuffer as a mutable reference. I wonder if we should make `toArray` effectful, too, to demonstrate it's capturing the current state of the ArrayBuffer.

jacereda replied:
Yes, I guess the right thing would be to make `toArray` effectful.

Oh, I switched right now to the v10 branch. I changed there the order of arguments for the `DataView` setters and the implementation. I don't like the `ViewTypeName` stuff there, maybe you can figure out some type-level black magic to avoid it?

athanclark replied:
Awesome, I like the new DataView. I'm not sure if I can get away with any typelevel magick here - purescript doesn't really have a type-level `class Show :: Type -> Symbol -> Constraint` typeclass in typelevel-prelude, so there's not really an automated way of printing out typenames. If there's a sneaky way of doing this, I'm all ears. But I think that might just have to wait for the next version of purs or something :\. The least I could do right now is omit the AProxy stuff from this typeclass - we'll just make it a type-level mapping, then reflect the type back to a string.

athanclark replied:
Alright, I've jerry-rigged a type-level Show for `ArrayViewType`-kinded types. I've also restricted Prelude imports compulsively - I'm planning on doing crazy stuff to the purs compiler down the line, and explicit imports always helps to determine cyclomatic complexity and what to inline.

jacereda replied:
Nice!

jacereda committed:
Make empty effectful

jacereda committed:
Restore specific setters/getters in the API, changed implementation

athanclark committed:
Merge branch 'v10' of git://github.com/jacereda/purescript-arraybuffer into redesign

athanclark committed:
fixing tests

athanclark committed:
restricted imports

athanclark commented:
Oooh, I just had a tricky thought... should offsets and byte length be represented as UInts? Because they can't be negative...? It feels against-the-grain though :\

jacereda commented:
In the manual bounds-checking they're interpreted as UInts... If there was compiler support so that literals could become UInts automagically it would probably be the way to go, but as it's now I guess it will just be burdensome.

athanclark commented:
Yeah I was thinking the same. We gotta get some kind of purescript v0.13 wish list going or something

jacereda commented on src/Data/ArrayBuffer/Typed.purs:
```diff
@@ -1,87 +1,431 @@
+
+-- | Returns a new typed array view of the same buffer, beginning at the index and ending at the second.
+-- |
+-- | **Note**: there is really peculiar behavior with `subArray` - if the first offset argument is omitted, or
```
Is this behaviour for `subarray` documented somewhere?

athanclark replied:
I wish. I can't find it in the ECMA spec, but it's pretty obvious if you try it in psci. I'm not sure how to find a doc on this either

jacereda commented on src/Data/ArrayBuffer/Typed.purs:
```diff
@@ -1,87 +1,431 @@
+foreign import joinImpl :: forall a. Fn2 (ArrayView a) String String
+
+-- | Prints array to a delimiter-separated string - see [MDN's spec](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/join) for details.
+toString' :: forall a. ArrayView a -> String -> String
```
`join`? `toStringSep`? `toStringSep`? `toStringIntercalate`?

athanclark replied:
I'll go with join :) sorry

jacereda commented:
I've pushed more changes to the v10 branch making more functions effectful. I guess some more would need to be converted, particularly I guess `at` should also be effectful, right?

athanclark commented:
Definitely, `at` is just `Ref.read` in a way

jacereda commented:
I've unified the signatures for `slice`/`subArray`/`fill` to mirror Array/Node.Buffer APIs in the v10 branch. Also, I can't reproduce the behaviour you specify for `subArray` and removed the note. I seem to remember fixing one of the `subArray` tests, maybe that was the problem (or maybe you found a broken node version?).

athanclark commented:
:s shoot I don't think so, but I wouldn't doubt it. iirc I was using nvm v8.12.0, but either way I'm glad the mystery is cleared up.

But awesome, I definitely should have thought of that - the API mirroring.

jacereda commented on src/Data/ArrayBuffer/Typed.js:
```diff
@@ -1,75 +1,185 @@
-exports.asInt8Array = function(v) {
-  return new Int8Array(v.buffer, v.byteOffset, v.byteLength);
-}
+exports.polyFill = function polyFill () {
```
Why do you export this? Would it be problematic to just ejecute the polyfill directly?

athanclark replied:
I don't think it would, but I was just trying to be on the safe side - it technically causes an effect, and although I know ffi files only get executed once, I just felt like someone might want finer granularity of control. But honestly I think it would be fine to just execute - I pulled it [from MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray#Methods_Polyfill).

jacereda commented:
I've added some unsafe instances (Show/Eq/Ord/Semigroup/Monoid), but I don't think these belong in this repo, I guess they could be somewhere else. See https://github.com/jacereda/purescript-arraybuffer/pull/24

athanclark commented:
Beautiful! I think it looks great. The newtype was exactly what I was going to suggest.

jacereda commented:
I've merged to master but will hold the release until I get comments from @AlexaDeWit regarding the new API. 

On a different subject, do you think TypedArrays could have a Foldable instance? I've been unable to write it...

athanclark commented:
Awesome, that will be great. I think a `cereal` compatible library is next on my plate after that. But technically, no, because it's not a container over "any type `a`"; only the type that it's related to - like a `Uint` or whathaveyou; it's got that extra constraint predicate.

jacereda commented:
But somehow it's a container for "some type `a`". Is it possible to specify that `a` could be any of [UInt,Int,Float32,...]?

athanclark commented:
No, because an instance of `Foldable` only exposes it's container type `f` for an instance constraint. Because `Foldable.foldr` adds an additional `forall a.` in it's member definition - there's no way to link that skolemized `a` with the `f` at the top level.

athanclark commented:
It's almost the same idea for why a `Set` can't be a `Functor` - because we have no way to say that the member elements must also have an `Ord` constraint.

jacereda commented:
> Awesome, that will be great. I think a `cereal` compatible library is next on my plate after that. But technically, no, because it's not a container over "any type `a`"; only the type that it's related to - like a `Uint` or whathaveyou; it's got that extra constraint predicate.

Not `cereal` compatible, but some time ago I started https://github.com/jacereda/purescript-binary and I guess I'll update it at some point to work with this.

athanclark commented:
Ping - hey sorry to pester. Were you intending on publishing the new changes in master? I know the `v9.0.0` version doesn't include the `Gen` modules and all of that. I'm just trying to publish some other libraries that rely on this one :) thanks for your help!

jacereda commented:
I was just waiting to see if @AlexaDeWit had any comment regarding the new API. In any case, I'll try to release on Thursday.

athanclark commented:
Awesome, thank you!

AlexaDeWit commented:
Sorry about that, I didn't mean to ignore things. I've just not been following this super closely, as legacy versions will work for me, and I kind of trust you two to come up with something coherent so that when I get around to making my codec lib it should be fine
